README
GEOCLUE SHARE
=============

This is the repository of an Android application named GeoclueShare.
GeoclueShare shares location information from android GPS to Geoclue [1].

This application is a part of GSoC 2015 project [2].

REQUIREMENTS FOR DEVELOPMENT
============================

 * Java Development Kit (JDK) 7 or higher.
 * Android Studio.
 * Android SDK with build tools and support libraries installed.
 * Android SDK build tools version 21.1.1 and target API 21 preferred.
 * Android device with GPS capabilities and running android version 2.2 or
   higher (for testing).

REFERENCES
==========

 [1] https://wiki.freedesktop.org/www/Software/GeoClue/
 [2] https://wiki.gnome.org/Outreach/SummerOfCode/2015/Projects/Ankit_GPSSharing

NEWS
v1.1
====

- Add polish translation.
- Bug fixes.
- Reject last known location older than 24 hours.

v1.0
====

- Get location from listener too, instead of just relying on NmeaListener.
- Send Geoclue the last known location from Android GPS so that the user might
  not have to wait longer.
- Add notification for the background service.
- Bug fixes.

v1.0-beta
=========

- First release (of GeoclueShare)



app/src/main/java/org/freedesktop/geoclueshare/Zeroconf.java
package org.freedesktop.geoclueshare;

import android.content.Context;
import android.net.wifi.WifiManager;
import android.text.format.Formatter;
import android.util.Log;

import java.io.IOException;
import java.net.InetAddress;
import java.util.HashMap;

import javax.jmdns.JmDNS;
import javax.jmdns.ServiceInfo;

/*
 * Copyright (C) 2015 Ankit (Verma)
 *
 * GeoclueShare is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * GeoclueShare is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along
 * with GeoclueShare; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * Author: Ankit (Verma) <ankitstarski@gmail.com>
 */

public class Zeroconf {
    private static final String TAG = "Zeroconf";
    private static WifiManager.MulticastLock multicastLock;
    private JmDNS jmdns;
    private ServiceInfo serviceInfo;
    private static String ip;

    /**
     * The default tag string for Multicast, used in {@link Zeroconf#attainLock(Context)}.
     */
    private static final String DEFAULT_MULTICAST_TAG_STRING = "GeoclueShare";

    /**
     * This function is to be called before Using {@link Zeroconf} class is to be used. <br/>
     * This attains Multicast lock from WiFi service which is requred for mDNS broadcasting or
     * listening.
     *
     * @param context the context from where it is being called.
     */
    public static void attainLock(Context context) {
        WifiManager wifi = (WifiManager)context.getSystemService(Context.WIFI_SERVICE);
        ip = Formatter.formatIpAddress(wifi.getConnectionInfo().getIpAddress());

        multicastLock = wifi.createMulticastLock(DEFAULT_MULTICAST_TAG_STRING);
        multicastLock.setReferenceCounted(true);
        multicastLock.acquire();
    }

    /**
     * Same as {@link Zeroconf#attainLock(Context)} but Tag string for the lock can be set.
     *
     * @param context the context from where it is being called.
     * @param tag     a tag for the MulticastLock to identify it in debugging
     *                messages.
     */
    public static void attainLock(Context context, String tag) {
        WifiManager wifi = (WifiManager)context.getSystemService(Context.WIFI_SERVICE);
        ip = Formatter.formatIpAddress(wifi.getConnectionInfo().getIpAddress());

        multicastLock = wifi.createMulticastLock(tag);
        multicastLock.setReferenceCounted(true);
        multicastLock.acquire();
    }

    /**
     * Releases the lock attained by {@link Zeroconf#attainLock}.
     */
    public static void releaseLock() {
        if (multicastLock != null) {
            multicastLock.release();
            multicastLock = null;
        }
    }

    public Zeroconf() {
        try {
            jmdns = JmDNS.create(InetAddress.getByName(ip));
        } catch (IOException e) {
            Log.d(TAG, "Can't start mDNS service");
        }
    }

    /**
     * Broadcasts mDNS service, {@code "_nmea-0183._tcp.local."} at a particular port number.
     *
     * @param serviceName name of the service to be shoown to other devices.
     * @param port port number
     */
    public void broadcastService(String serviceName, int port) {
        try {

            /*
             * Beware of this bug in JmDNS
             * http://stackoverflow.com/questions/12726801/avahi-not-able-to-find-service-creted-by-jmdns
             */
            HashMap<String, byte[]> properties = new HashMap<String, byte[]>();
            String serviceText = "Location Server for Geoclue";
            properties.put("description", serviceText.getBytes());
            properties.put("accuracy", LocationService.accuracy.getBytes());

            if (serviceName == null || serviceName.length() == 0)
                serviceName = DEFAULT_MULTICAST_TAG_STRING;

            serviceInfo = ServiceInfo.create("_nmea-0183._tcp.local.",
                    serviceName, port, 0, 0, true,
                    properties);

            jmdns.registerService(serviceInfo);

        } catch (Exception e) {
            Log.d(TAG, "Can't register Service");
        }
    }

    /**
     * Unregisters the service being broadcasted by {@link Zeroconf#broadcastService}.
     */
    public void unregisterService() {
        if(serviceInfo != null) {
            try {
                jmdns.unregisterService(serviceInfo);
            } catch (NullPointerException e) {
                Log.w(TAG, "Can't close a null mDNS service.");
            }
        }
    }
}


app/src/main/java/org/freedesktop/geoclueshare/MainActivity.java
package org.freedesktop.geoclueshare;

import android.app.ActivityManager;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.provider.Settings;
import android.support.v7.app.ActionBarActivity;
import android.support.v7.widget.SwitchCompat;
import android.util.Log;
import android.widget.CompoundButton;
import android.widget.Switch;
import android.widget.TextView;

/*
 * Copyright (C) 2015 Ankit (Verma)
 *
 * GeoclueShare is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * GeoclueShare is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along
 * with GeoclueShare; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * Author: Ankit (Verma) <ankitstarski@gmail.com>
 */

public class MainActivity extends ActionBarActivity {

    private static final String TAG = "MainActivity";

    /**
     * The code used to depict Location Settings activity when prompting for High accuracy.
     */
    private static final int LOCATION_MODE_CHANGE = 0;

    /**
     * The message code to prompt for High accuracy location settings. This constant is supposed to
     * be used by {@link #handler}.
     */
    private static final int MESSAGE_PROMPT_LOCATION = 0;

    /**
     * The message code used by {@link #handler} in order to update the number of clients in the
     * UI.
     */
    private static final int MESSAGE_CONNECTED_DEVICES = 1;

    /**
     * The key string used by Handler data bundle for updating the number of clients to the
     * MainActivity's UI.
     */
    private static final String KEY_CONNECTED_DEVICES = "n_clients";


    private SwitchCompat toggleService;
    private TextView locationMode;
    private TextView connectedDevices;
    private static Handler handler;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        getSupportActionBar().setTitle(R.string.main_activity_label);
        final Intent locationServiceIntent = new Intent(this, LocationService.class);

        toggleService = (SwitchCompat) findViewById(R.id.service_toggle);
        locationMode = (TextView) findViewById(R.id.location_mode);
        connectedDevices = (TextView) findViewById(R.id.connected_devices);

        handler = new Handler(Looper.getMainLooper()) {
            @Override
            public void handleMessage(Message message) {
                switch (message.what) {
                    case MESSAGE_PROMPT_LOCATION:
                        showLocationPrompt();
                        Log.d(TAG, "Location Prompt");
                        break;
                    case MESSAGE_CONNECTED_DEVICES:
                        Bundle bundle = message.getData();
                        int n_clients = bundle.getInt(KEY_CONNECTED_DEVICES);
                        connectedDevices.setText(n_clients + "");
                        Log.d(TAG, "Connected devices changed");
                        break;
                }
            }
        };

        if (isLocationAccuracyHigh()) {
            locationMode.setText(R.string.location_high_accuracy);
        } else {
            locationMode.setText(R.string.location_low_accuracy);
            toggleService.setClickable(false);
            showLocationPrompt();
        }

        if (isServiceRunning(LocationService.class)) {
            connectedDevices.setText(NetworkListener.numberOfClients + "");
        }

        toggleService.setChecked(isServiceRunning(LocationService.class));

        toggleService.setOnCheckedChangeListener(new Switch.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                if (isChecked) {
                    startService(locationServiceIntent);
                } else {
                    stopService(locationServiceIntent);
                }
            }
        });
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode,
                                    Intent data) {
        if (requestCode == LOCATION_MODE_CHANGE) {
            if (isLocationAccuracyHigh()) {
                locationMode.setText(R.string.location_high_accuracy);
                toggleService.setClickable(true);
            } else {
                locationMode.setText(R.string.location_low_accuracy);
                showLocationPrompt();
            }
        }
    }

    private boolean isServiceRunning(Class<?> serviceClass) {
        ActivityManager manager = (ActivityManager) getSystemService(Context.ACTIVITY_SERVICE);
        for (ActivityManager.RunningServiceInfo service : manager.getRunningServices(Integer.MAX_VALUE)) {
            if (serviceClass.getName().equals(service.service.getClassName())) {
                return true;
            }
        }
        return false;
    }

    private boolean isLocationAccuracyHigh() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            try {
                int locationAccuracy = Settings.Secure.getInt(getContentResolver(),
                        Settings.Secure.LOCATION_MODE);
                if (locationAccuracy == Settings.Secure.LOCATION_MODE_HIGH_ACCURACY ||
                        locationAccuracy == Settings.Secure.LOCATION_MODE_SENSORS_ONLY) {

                    return true;
                }
            } catch (Settings.SettingNotFoundException e) {
                Log.d(TAG, "Error in detecting location settings");
            }
        } else {
            String locationProviders = Settings.Secure.getString(getContentResolver(),
                    Settings.Secure.LOCATION_PROVIDERS_ALLOWED);
            return locationProviders.contains("gps");
        }

        return false;
    }

    private void showLocationPrompt() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setMessage(R.string.location_prompt_dialog)
                .setPositiveButton(R.string.yes, new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        startActivityForResult(new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS),
                                LOCATION_MODE_CHANGE);
                    }
                })
                .setNegativeButton(R.string.no, new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        // User cancelled the dialog
                    }
                });

        builder.show();
    }

    public static void promptForLocation() {
        Message message = handler.obtainMessage(MESSAGE_PROMPT_LOCATION);
        message.sendToTarget();
    }

    public static void setConnectedDevices(int n_clients) {
        Bundle bundle = new Bundle();
        Message message = handler.obtainMessage(MESSAGE_CONNECTED_DEVICES);
        bundle.putInt(KEY_CONNECTED_DEVICES, n_clients);
        message.setData(bundle);
        message.sendToTarget();
    }
}


app/src/main/java/org/freedesktop/geoclueshare/NetworkListener.java
package org.freedesktop.geoclueshare;

import android.os.AsyncTask;
import android.os.Handler;
import android.os.Message;
import android.util.Log;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.SelectionKey;
import java.nio.channels.Selector;
import java.nio.channels.ServerSocketChannel;
import java.nio.channels.SocketChannel;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;

/*
 * Copyright (C) 2015 Ankit (Verma)
 *
 * GeoclueShare is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * GeoclueShare is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along
 * with GeoclueShare; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * Author: Ankit (Verma) <ankitstarski@gmail.com>
 */

/**
 * {@code NetworkListener} creates a new Thread (using AsyncTask) for network tasks. It performs
 * tasks such as client connection, disconnection, send or receive data.
 * <p>Call {@link NetworkListener#sendData} to broadcast data to all the clients.</p>
 */
public class NetworkListener extends AsyncTask<Void, Void, Void> {

    private static final String TAG = "NetworkListener";
    private ServerSocketChannel server = null;
    private Handler handler;
    private static HashMap<String, String> pendingData;
    Zeroconf mdns;

    /**
     * Total number of clients currently connected to the application.
     */
    public static int numberOfClients = 0;

    /**
     * The TCP/IP port used for Socket communication.
     */
    private static final int PORT = 10110;

    /**
     * The maximum length of a GGA sentence.
     */
    private static final int GGA_LENGTH_MAX = 80;

    public NetworkListener(Handler handler) {
        this.handler = handler;
    }

    @Override
    protected Void doInBackground(Void... params) {
        Selector selector;
        Iterator i;

        Log.d(TAG, "Started Listening");

        pendingData = new HashMap<String, String>();

        try {
            startServer();

            mdns = new Zeroconf();
            mdns.broadcastService(LocationService.deviceId, PORT);

            selector = Selector.open();
            server.register(selector, SelectionKey.OP_ACCEPT);
        } catch (IOException e) {
            Log.d(TAG, "Unable to create ServerSocket for port: " + PORT);
            Log.d(TAG, e.getMessage());
            return null;
        }

        while (!isCancelled()) {
            try {
                selector.select();
                Set keys = selector.selectedKeys();
                i = keys.iterator();
            } catch (IOException e) {
                Log.d(TAG, "Selector can't select");
                continue;
            }

            while (i.hasNext()) {
                SelectionKey key = (SelectionKey) i.next();
                i.remove();

                if (isCancelled()) {
                    return null;
                }

                try {
                    if (key.isAcceptable()) {
                        SocketChannel client = server.accept();
                        addClient(client, selector);

                        continue;
                    }

                    if (key.isReadable()) {
                        SocketChannel client = (SocketChannel) key.channel();
                        disconnectClient(client);

                        continue;
                    }

                    if (key.isWritable()) {
                        SocketChannel client = (SocketChannel) key.channel();
                        sendDataToClient(client);
                    }
                } catch (IOException e) {
                    continue;
                }
            }
        }

        return null;
    }

    /**
     * Broadcasts data to all the connected clients at port number stored inside
     * {@link NetworkListener#PORT}
     *
     * @param data The data that is to be broadcasted to all the connected clients at port number
     *             stored inside {@link NetworkListener#PORT}.
     */
    public static void sendData(String data) {
        String ggaSentence = data + "\r\n";

        Iterator it = pendingData.entrySet().iterator();
        while (it.hasNext()) {
            Map.Entry pair = (Map.Entry) it.next();
            pair.setValue(ggaSentence);
        }
    }

    private void startServer() throws IOException {
        server = ServerSocketChannel.open();
        server.configureBlocking(false);
        server.socket().bind(new java.net.InetSocketAddress(PORT));
    }

    private void addClient(SocketChannel client, Selector selector) throws IOException {
        client.configureBlocking(false);
        client.register(selector, SelectionKey.OP_READ | SelectionKey.OP_WRITE);
        pendingData.put(client.socket().getInetAddress().toString(), "");

        Log.d(TAG, "Client connected");
        numberOfClients++;
        MainActivity.setConnectedDevices(numberOfClients);

        Log.d(TAG, "Number of clients: " + numberOfClients);
        if (numberOfClients == 1) {
            Message message = handler.obtainMessage(
                    LocationService.MESSAGE_START_GPS);
            message.sendToTarget();
        }
    }

    private void disconnectClient(SocketChannel client) throws IOException {
        int BUFFER_SIZE = 32;
        ByteBuffer buffer = ByteBuffer.allocate(BUFFER_SIZE);

        try {
            if (client.read(buffer) == -1) {
                throw new Exception();
            }
        } catch (Exception e) {
            pendingData.remove(client.socket().getInetAddress().toString());
            client.close();
            Log.d(TAG, "Client disconnected");

            numberOfClients--;
            MainActivity.setConnectedDevices(numberOfClients);

            Log.d(TAG, "Number of clients: " + numberOfClients);
            if (numberOfClients == 0) {
                Message message = handler.obtainMessage(
                        LocationService.MESSAGE_STOP_GPS);
                message.sendToTarget();
            }
        }
    }

    private void sendDataToClient(SocketChannel client) throws IOException {
        String data = pendingData.get(client.socket().getInetAddress().toString());

        if (data == null || data.length() == 0)
            return;

        ByteBuffer buf = ByteBuffer.allocate(GGA_LENGTH_MAX);
        buf.clear();

        buf.put(data.getBytes());
        buf.flip();

        while (buf.hasRemaining()) {
            client.write(buf);
        }

        pendingData.put(client.socket().getInetAddress().toString(), "");
    }

    @Override
    protected void onCancelled() {
        super.onCancelled();
        try {
            if (mdns != null)
                mdns.unregisterService();
            if (server != null)
                server.close();
            numberOfClients = 0;
            MainActivity.setConnectedDevices(numberOfClients);

            if (pendingData != null)
                pendingData.clear();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}


app/src/main/java/org/freedesktop/geoclueshare/LocationService.java
package org.freedesktop.geoclueshare;

import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Intent;
import android.content.ServiceConnection;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.location.GpsStatus;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.os.Message;
import android.provider.Settings;
import android.support.annotation.Nullable;
import android.support.v4.app.NotificationCompat;
import android.util.Log;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.TimeZone;

/*
 * Copyright (C) 2015 Ankit (Verma)
 *
 * GeoclueShare is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * GeoclueShare is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along
 * with GeoclueShare; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * Author: Ankit (Verma) <ankitstarski@gmail.com>
 */

/**
 * <p>
 * {@code LocationService} is responsible for fetching and handling location data from the GPS in
 * the background.
 * </p>
 * FIXME: It's just a dummy class yet. Add functionality.
 */
public class LocationService extends Service implements LocationListener, GpsStatus.NmeaListener {

    private static final String TAG = "LocationService";
    private LocationManager locationManager;
    private NetworkListener networkListener;
    private String ggaSentence;
    private NotificationCompat.Builder builder;

    /**
     * The unique identifier for current Android device.
     */
    public static String deviceId;

    /**
     * The that value goes into the `accuracy` feild of mDNS service's TXT record.
     */
    public static String accuracy = "exact";

    /**
     * The minimum time between updates in milliseconds.
     */
    private static final long MIN_TIME_BW_UPDATES_GPS = 1000;

    /**
     * The minimum distance to change Updates in meters.
     */
    private static final long MIN_DISTANCE_CHANGE_FOR_UPDATES = 1;

    /**
     * The code for stopping Location updates.
     */
    public static final int MESSAGE_STOP_GPS = 0;

    /**
     * The code for starting Location updates.
     */
    public static final int MESSAGE_START_GPS = 1;

    /**
     * The max allowed age of last known location.
     */
    public static final int LAST_KNOWN_LOCATION_MAX_AGE = 24 * 60 * 60 * 1000;

    /**
     * Notification id.
     */
    private static final int NOTIFICATION_ID = 007;

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.d(TAG, "Service started");
        createNotification();

        Zeroconf.attainLock(this);

        locationManager = (LocationManager) getSystemService(LOCATION_SERVICE);

        deviceId = Settings.Secure.getString(getApplicationContext().getContentResolver(),
                Settings.Secure.ANDROID_ID);

        Handler handler = new Handler(Looper.getMainLooper()) {
            @Override
            public void handleMessage(Message message) {
                switch (message.what) {
                    case MESSAGE_START_GPS:
                        startGps();
                        Log.d(TAG, "GPS start");
                        Location loc = locationManager
                                .getLastKnownLocation(LocationManager.GPS_PROVIDER);

                        if (loc != null && System.currentTimeMillis() - loc.getTime() >
                                LAST_KNOWN_LOCATION_MAX_AGE)
                            loc = null;

                        if (loc != null)
                            NetworkListener.sendData(getGgaFromLocation(loc));
                        break;
                    case MESSAGE_STOP_GPS:
                        stopGps();
                        Log.d(TAG, "GPS stop");
                }
            }
        };

        networkListener = new NetworkListener(handler);
        networkListener.execute();

        return Service.START_STICKY;
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void unbindService(ServiceConnection conn) {
        super.unbindService(conn);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        stopGps();
        networkListener.cancel(true);
        Zeroconf.releaseLock();

        Log.d(TAG, "Service destroyed");
        removeNotification();
    }

    @Override
    public void onNmeaReceived(long timestamp, String nmea) {
        String nmeaTimestamp = nmea.split(",")[1];
        if (nmea.startsWith("$GPGGA") && nmeaTimestamp.matches("[0-9]{6}")) {
            NetworkListener.sendData(nmea);
            ggaSentence = null;
        } else if (nmea.startsWith("$GPGGA")) {
            if (ggaSentence == null || ggaSentence.length() == 0)
                return;
            nmeaTimestamp = ggaSentence.split(",")[1];

            if (!nmeaTimestamp.matches("[0-9]{6}"))
                return;
            NetworkListener.sendData(ggaSentence);
            ggaSentence = null;
        }
    }

    private void startGps() {
        locationManager.requestLocationUpdates(
                LocationManager.GPS_PROVIDER,
                MIN_TIME_BW_UPDATES_GPS,
                MIN_DISTANCE_CHANGE_FOR_UPDATES,
                this);
        locationManager.addNmeaListener(this);
    }

    private void stopGps() {
        locationManager.removeUpdates(this);
        locationManager.removeNmeaListener(this);
    }

    @Override
    public void onLocationChanged(Location location) {
        ggaSentence = getGgaFromLocation(location);
    }

    @Override
    public void onStatusChanged(String provider, int status, Bundle extras) {

    }

    @Override
    public void onProviderEnabled(String provider) {

    }

    @Override
    public void onProviderDisabled(String provider) {
        if (provider.equals(LocationManager.GPS_PROVIDER)) {
            MainActivity.promptForLocation();
        }
    }

    private String getGgaFromLocation(Location location) {
        String gga;
        Boolean hasAltitude = location.hasAltitude();

        Date date = new Date(location.getTime());
        DateFormat format = new SimpleDateFormat("HHmmss");
        format.setTimeZone(TimeZone.getTimeZone("UTC"));
        String time = format.format(date);

        if (hasAltitude) {
            gga = "$GPGGA,%s,%s,%s,1,,%.1f,%.1f,M,,M,,";
            gga = String.format(gga,
                    time,
                    getLatitudeString(location.getLatitude()),
                    getLongitudeString(location.getLongitude()),
                    getHdopFromAccuracy(location.getAccuracy()),
                    location.getAltitude());
        } else {
            gga = "$GPGGA,%s,%s,%s,1,,%.1f,,M,,M,,";
            gga = String.format(gga,
                    time,
                    getLatitudeString(location.getLatitude()),
                    getLongitudeString(location.getLongitude()),
                    getHdopFromAccuracy(location.getAccuracy()));
        }

        gga = addChecksumToGga(gga);

        return gga;
    }

    private double getHdopFromAccuracy(double accuracy) {
        /* FIXME: These are rough estimates based on the information given in the link below:
         *        http://en.wikipedia.org/wiki/Dilution_of_precision_%28GPS%29#Meaning_of_DOP_Values
         */
        if (accuracy <= 0.5)
            return 0.5;
        else if (accuracy <= 1.0)
            return 1.5;
        else if (accuracy <= 3.0)
            return 3.5;
        else if (accuracy <= 50.0)
            return 7.5;
        else if (accuracy <= 100.0)
            return 15.0;
        else
            return 30.0;
    }

    private String getLatitudeString(double lat) {
        String latStr = "%02d%06.3f,%s";

        int degrees = (int) Math.abs(lat);
        double minutes = Math.abs((lat - (int) lat) * 60);
        String symbol = lat >= 0 ? "N" : "S";

        latStr = String.format(latStr, degrees, minutes, symbol);
        return latStr;
    }

    private String getLongitudeString(double lon) {
        String lonStr = "%03d%06.3f,%s";

        int degrees = (int) Math.abs(lon);
        double minutes = Math.abs((lon - (int) lon) * 60);
        String symbol = lon >= 0 ? "E" : "W";

        lonStr = String.format(lonStr, degrees, minutes, symbol);
        return lonStr;
    }

    private String addChecksumToGga(String gga) {
        int checksum = 0;

        for (int i = 1, n = gga.length(); i < n; i++) {
            checksum ^= (int) gga.charAt(i);
        }

        return String.format("%s*%02X", gga, checksum);
    }

    private void createNotification() {
        Bitmap icon = BitmapFactory.decodeResource(getResources(), R.mipmap.ic_launcher);

        builder = new NotificationCompat.Builder(this)
                .setSmallIcon(R.mipmap.ic_location_share)
                .setLargeIcon(icon)
                .setContentTitle(getString(R.string.notif_title))
                .setContentText(getString(R.string.notif_content))
                .setOngoing(true);

        Intent resultIntent = new Intent(this, MainActivity.class);
        resultIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_SINGLE_TOP);

        PendingIntent resultPendingIntent = PendingIntent.getActivity(
                this,
                0,
                resultIntent,
                PendingIntent.FLAG_UPDATE_CURRENT
        );
        builder.setContentIntent(resultPendingIntent);

        NotificationManager notoficationManager =
                (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
        notoficationManager.notify(NOTIFICATION_ID, builder.build());
    }

    private void removeNotification() {
        NotificationManager notoficationManager =
                (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
        notoficationManager.cancel(NOTIFICATION_ID);
    }
}
